Fix control flow around bucket overflows in make_pmh_buckets

garbageslam opened the pull request:
This is a correctness issue reported in frozen issue #95

The issue is that the `continue` is meant to continue the outer
`while (1)` loop, so that we wipe out the buckets and try another
seed. As written, the code will simply not place the item that
overflowed. Then lookup of that item will likely fail.

I have never seen this happen in practice, I think because we
make the buckets pretty large, like sqrt(N), during the construction.
So likely overflow never occurs.
Otherwise I would expect the map to give a junk answer in practice.

I think I introduced this bug in this commit:
https://github.com/serge-sans-paille/frozen/commit/2897706fc23936a04666e9e9e3d60903c7866c7b#diff-b7220304cb6955468f4a0071ade6af7e

jibal commented:
Since the purpose of original change was never realized, you might want to reconsider the whole approach. It seems to me that limiting the size of the buckets is not a good idea ... you actually want to maximize the number of buckets of size 1, which increases the speed of lookup. That also means that H can be smaller--there's no reason for the size of H to be M, the bucket count ... its size actually ought to be around number_of_items - number_of_singletons. In fact, making the size of H be M may make it too small to possibly hold all the items.

garbageslam commented:
> Since the purpose of the original change was never realized, you might want to reconsider the whole approach. It seems to me that limiting the size of the buckets is not a good idea ... you actually want to maximize the number of buckets of size 1, which increases the speed of lookup.

I think you have a misunderstanding about the algorithm. These buckets only exist during PMH creation. The operation of the algorithm at runtime is always exactly two table lookups, one in the "first phase" hash table, and one in the "second phase" hash table. It is constant time regardless of the "bucket sizes", the buckets only exist when we are deciding exactly what data to write in the first hash table entry.

Limiting the size of the buckets is crucial, and it fixed problems in the library during development two years ago.

In the second phase of the algorithm, after the buckets have been created, they are sorted and considered, largest first. For each one we try by random guessing to find a seed to a hash function that places all those items into the second table without collisions. We try seeds over and over until we find one such that none of them collide and none of them hit anything that has already been placed.

The expected time to find a seed that works for any given bucket is ~ `(1-load- (bucket_size / second table size))^bucket_size` if we model the seeded hash function as an ideal random function.
If load is a constant, this is exponential in bucket size, so, very bad if there are large buckets.
This means that if the first hash is bad and many items are in one bucket, then phase 2 will fail or become an infinite loop.

2 years ago, it was common that you would try to create the hash for some strings and ints or whatever, and compilation fails with "recursion limit reached". My recollection is that we fixed this by improving the hash functions, and also by having this check in phase 1 that tries to redo the first round hash if the first round hash was bad. That's what the bucket size limit is doing. We also adjusted the size of the buckets.

For more background I would suggest to study FKS perfect hashing which is where all of these ideas came from. This scheme used in frozen is only a variation on that, and I think it was initially developed by Serge, I'm not sure exactly if he saw it somewhere else or made it up. FKS perfect hashing also only works if the first-round hash is good, and if it's not, the best thing to do is retry, because the second round cannot succeed.

> It seems to me that limiting the size of the buckets is not a good idea ... you actually want to maximize the number of buckets of size 1, which increases the speed of lookup. That also means that H can be smaller--there's no reason for the size of H to be M, the bucket count ... its size actually ought to be around number_of_items - number_of_singletons. In fact, making the size of H be M may make it too small to possibly hold all the items.

Yes, you would love to maximize the number of buckets of size 1. That is the entire goal in perfect hashing -- to make every bucket of size 1. But it is impossible to do this in the "first round" with a conventional hash function. If you can make all the buckets of size 1 in one round, then you don't need a two-phase algorithm at all. 

It is known for information-theoretic reasons that the description length of a perfect hash table is generally `O(n)` if there are `n` items. But in this code, the first-round hash table has like a 64 byte seed no matter how many items there are. So its description length is O(1). So that hash cannot be perfect. This is why you need the two-phase thing -- the extra description complexity for the hash function is essentially what is held in the "first round" table. This is also the genius of the FKS idea -- they observed that as long as you have merely pairwise-independent hash function families, you can build perfect hashing by a two phase process, where in the first phase you make at least decent progress partitioning the items, and then in the second phase you hash the parts of that partition into a second arena in expected polynomial time.

The reasoning behind the bucket size limit is, even if there are a lot of buckets of size 1, if e.g. half the items are in 1/epsilon of the buckets, it might take 2^{1/epsilon} time during phase 2. An easy way to filter this out is to reject large buckets. But a more refined criteria may be better -- it is probably okay if the largest bucket is "fairly big" and all the other buckets are small. The running time was studied in detail in FKS and subsequent papers and there are much more precise estimates known, but this is simple and seems to work in practice.

> In fact, making the size of H be M may make it too small to possibly hold all the items.

It is known that a perfect hash table for `N` items has size `O(N)` in general, and this `O` is like 1 + epsilon, not less than one.

In my recollection, in this code, `M` is the number of slots in the second table, and `M` is like `2N` or something, in part because it's better from hardware point of view if that table has size which is a power of two, so you don't need `IDIV` instructions etc.

Maybe the present revision of the library is different? I would have to review the code to be sure.

But if the library is currently configured to choose M < N then there is no way that it could be working at all for any input, so that should not be the case if the tests are passing.

garbageslam commented:
I guess that we are still doing this:

```
template <class Key, class Value, std::size_t N, typename Hash = anna<Key>,
          class KeyEqual = std::equal_to<Key>>
class unordered_map {
  static constexpr std::size_t storage_size =
      bits::next_highest_power_of_two(N) * (N < 32 ? 2 : 1); // size adjustment to prevent high collision rate for small sets
  using container_type = bits::carray<std::pair<Key, Value>, N>;
  using tables_type = bits::pmh_tables<storage_size, Hash>;
```
https://github.com/serge-sans-paille/frozen/blob/master/include/frozen/unordered_map.h#L53

So the "storage size" `M` is `N` rounded up to a power of two, then doubled again for small sets.

And we have two tables of size `M`, the first round one, and the second round one.

```
// Represents the perfect hash function created by pmh algorithm
template <std::size_t M, class Hasher>
struct pmh_tables {
  uint64_t first_seed_;
  carray<seed_or_index, M> first_table_;
  carray<std::size_t, M> second_table_;
  Hasher hash_;
```
https://github.com/serge-sans-paille/frozen/blob/master/include/frozen/bits/pmh.h#L149

So, for large `N`, I think we are probably within a factor 2 or so of the theoretical optimium.

Actually, I'm a little surprised to see this code on master. I think conventionally the second table would be like `carray<pair<Key, Value>, M>` rather than `carray<size_t, M>` to save a third round of indirection. But it requires more templates to propagate through the PMH code, so maybe that's why we never did it?

Don't get me wrong, I think there are still a lot of ways that this can be improved. There were a bunch of things I wanted to do with this to try to optimize it, but I changed jobs, now I work in rust, and I don't really use this library anymore. But I would caution against removing the bucket size limit.

Actually, if the logic around the bucket size limit was broken, and that wasn't causing failures in practice, that may mean that we should reduce the bucket size limit to make compilation a little faster.

I think this is where it is currently set:

```
  // TODO: Come up with justification for this, should it not be O(log M)?
  static constexpr auto bucket_max = 2 * (1u << (log(M) / 2));
```

https://github.com/serge-sans-paille/frozen/blob/master/include/frozen/bits/pmh.h#L54

jibal commented:
> I think you have a misunderstanding about the algorithm.

You're mistaken. I've now implemented this in Nim, incorporating my ideas as expressed above (however, I did limit the bucket size because with overly large buckets its too hard to distribute the bucket into H without collisions). I made the size of G and H independent, and am able to get smaller tables than your code generates ... and I don't need to provide M as an argument; the code always calculates optimal sizes for G and H (within limits imposed by retry thresholds). It runs either at compile-time or runtime because Nim is extremely flexible--and the code is easy to read because compile-time code looks exactly like runtime code in Nim.

I'm sorry that you started with a faulty ad hominem about my level of understanding, but it's your loss. Unsubscribing.

garbageslam commented:
> It is your understanding of the algorithm that is mistaken. When G (first_table) contains a singleton (resulting from a bucket with only one element, with isSeed false), there is only one lookup.

Here's the current code for lookup.

```
  // Looks up a given key, to find its expected index in carray<Item, N>
  // Always returns a valid index, must use KeyEqual test after to confirm.
  template <typename KeyType>
  constexpr std::size_t lookup(const KeyType & key) const {
    auto const d = first_table_[hash_(key, static_cast<size_t>(first_seed_)) % M];
    if (!d.is_seed()) { return static_cast<std::size_t>(d.value()); } // this is narrowing uint64 -> size_t but should be fine
    else { return second_table_[hash_(key, static_cast<std::size_t>(d.value())) % M]; }
  }
```

https://github.com/serge-sans-paille/frozen/blob/master/include/frozen/bits/pmh.h#L54

I guess that the code has changed in the last year or so, I don't think it was this way while I was actively contributing. Sorry for the mistake.

It would likely be better IMO if lookup returned the value and all the values were in the second table, so that there are never 3 levels of indirection through a pointer to get to the value. But perhaps for large values it's less memory overall not to do that.

> I made the size of G and H independent, and am able to get smaller tables than your code generates

Awesome, good job. There are surely a lot of optimizations that could be done here to make smaller tables.

> however, I did limit the bucket size because with overly large buckets its too hard to distribute them into H without collisions

:) yup that's my point

>> Limiting the size of the buckets is crucial, and it fixed problems in the library during development two years ago.
> It could not have done so, because of the bug that I identified and that was corrected here.

The order of events was roughly, as I very dimly remember, initially we had fairly weak hash functions, and Serge worked on finding better ones. In the meantime I developed the patch with the retries for a bad first hash. That made a difference when the hash function wasn't so great. With the current hash functions it may not make a difference, but that doesn't mean it should be removed IMHO.

> Um, it's clearly stated where it came from: http://stevehanov.ca/blog/?id=119 (which I read before discovering this repository).

I do remember reading that paper now, thanks.

> I'm sorry that you started with a faulty ad hominem about my level of understanding 

I'm sorry that you feel that you have been attacked. It was my intention merely to provide context for how I think this code came to be the way it is, and to explain why I don't think we should remove the bucket size limit. In engineering it is necessary to be able to speak against an idea without this being interpretted as a personal attack. It's also necessary to be able to point out point of disagreement in order to help reach shared understanding and make progress. I'm sorry if my tone did not come across well.  Thanks for contributing to the library.

serge-sans-paille commented:
Leaving aside the other part of the conversation, everyone agrees on the fix. Thanks @garbageslam and @jibal for the buf report and patch. And the interesting discussion thread.

As a side note, I indeed invited nothing with this lib, just porting the existing algorithm from  http://stevehanov.ca/blog/?id=119 to a constexpr form.
